File: app/src/main/java/com/clockmods/sdk/clock/WorldClockEntry.java
package com.clockmods.sdk.clock;

import java.util.TimeZone;

/** Immutable city/time-zone data supplied to styles that render a world-clock strip. */
public final class WorldClockEntry {
    private final String id;
    private final String city;
    private final String country;
    private final String zoneId;
    private final String flag;

    public WorldClockEntry(String id, String city, String country, String zoneId, String flag) {
        this.id = required(id, "id");
        this.city = required(city, "city");
        this.country = clean(country);
        this.zoneId = required(zoneId, "zoneId");
        this.flag = clean(flag);
        if (TimeZone.getTimeZone(this.zoneId).getID().equals("GMT")
                && !"GMT".equalsIgnoreCase(this.zoneId)) {
            throw new IllegalArgumentException("Unknown time zone: " + zoneId);
        }
    }

    public String getId() { return id; }
    public String getCity() { return city; }
    public String getCountry() { return country; }
    public String getZoneId() { return zoneId; }
    public String getFlag() { return flag; }

    /** Converts an ISO 3166-1 alpha-2 code to its Unicode regional-indicator flag. */
    public String getFlagEmoji() {
        if (flag.length() != 2) return flag;
        String upper = flag.toUpperCase(java.util.Locale.ROOT);
        char first = upper.charAt(0);
        char second = upper.charAt(1);
        if (first < 'A' || first > 'Z' || second < 'A' || second > 'Z') return flag;
        return new String(Character.toChars(0x1F1E6 + first - 'A'))
                + new String(Character.toChars(0x1F1E6 + second - 'A'));
    }

    private static String required(String value, String label) {
        String result = clean(value);
        if (result.length() == 0) throw new IllegalArgumentException(label + " must not be blank");
        return result;
    }

    private static String clean(String value) { return value == null ? "" : value.trim(); }

    @Override public boolean equals(Object other) {
        return other instanceof WorldClockEntry && id.equals(((WorldClockEntry) other).id);
    }

    @Override public int hashCode() { return id.hashCode(); }
}

File: app/src/main/java/com/clockmods/sdk/clock/WorldClockEntry.kt
package com.clockmods.sdk.clock

import java.util.Locale
import java.util.TimeZone

/** Immutable city/time-zone data supplied to styles rendering a world-clock strip. */
class WorldClockEntry(
    id: String?,
    city: String?,
    country: String?,
    zoneId: String?,
    flag: String?,
) {
    private val id = required(id, "id")
    private val city = required(city, "city")
    private val country = clean(country)
    private val zoneId = required(zoneId, "zoneId")
    private val flag = clean(flag)

    init {
        if (TimeZone.getTimeZone(this.zoneId).id == "GMT" &&
            !this.zoneId.equals("GMT", ignoreCase = true)
        ) {
            throw IllegalArgumentException("Unknown time zone: $zoneId")
        }
    }

    fun getId() = id
    fun getCity() = city
    fun getCountry() = country
    fun getZoneId() = zoneId
    fun getFlag() = flag

    fun getFlagEmoji(): String {
        if (flag.isEmpty()) return "\uD83C\uDF10"
        if (flag.length != 2) return flag
        val upper = flag.uppercase(Locale.ROOT)
        val first = upper[0]
        val second = upper[1]
        if (first !in 'A'..'Z' || second !in 'A'..'Z') return flag
        return String(Character.toChars(0x1F1E6 + first.code - 'A'.code)) +
            String(Character.toChars(0x1F1E6 + second.code - 'A'.code))
    }

    override fun equals(other: Any?) = other is WorldClockEntry && id == other.id
    override fun hashCode() = id.hashCode()

    private companion object {
        fun clean(value: String?) = value?.trim().orEmpty()
        fun required(value: String?, label: String) = clean(value).also {
            if (it.isEmpty()) throw IllegalArgumentException("$label must not be blank")
        }
    }
}
